### PR TITLE
Allow to assign sdw firmware name from default_fw_filename

### PR DESCRIPTION
**Pull request by libinyang**
The old code always uses sof_fw_filename in struct snd_soc_acpi_mach as the firmware name. This PR allows to use firmware name from default_fw_filename.

### CODE REVIEW

**Comment by libinyang** [comment]
@kv2019i I'm not sure whether ADL-S should apply the similar patch as tgl-h. The default fw filename is sof-adl-s.ri and ADL-S SDW is using sof-adl.ri.

File: sound/soc/sof/intel/hda.c [hunk 1]
```diff
@@ -1173,7 +1173,10 @@ static int hda_sdw_machine_select(struct snd_sof_dev *sdev)
 			mach->mach_params.links = mach->links;
 			mach->mach_params.link_mask = mach->link_mask;
 			mach->mach_params.platform = dev_name(sdev->dev);
-			pdata->fw_filename = mach->sof_fw_filename;
+			if (mach->sof_fw_filename)
+				pdata->fw_filename = mach->sof_fw_filename;
+			else
+				pdata->fw_filename = pdata->desc->default_fw_filename;
```
Comment (lyakh) [inline sound/soc/sof/intel/hda.c:1179]: using a gcc extension this could be
```
pdata->fw_filename = mach->sof_fw_filename ? : pdata->desc->default_fw_filename;
```

**Review by bardliao — DISMISSED** [review]
LGTM. I am fine with both `if/else` and `? :` style

File: sound/soc/sof/intel/hda.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda.c:1179]: I think the sound maintainers esp Mark prefer the if/else

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 2]
```diff
@@ -352,7 +352,6 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
 		.link_mask = 0xF, /* 4 active links required */
 		.links = tgl_3_in_1_default,
 		.drv_name = "sof_sdw",
-		.sof_fw_filename = "sof-tgl.ri",
```
Comment (lyakh) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c:355]: what about other entries in the `snd_soc_acpi_intel_tgl_sdw_machines[]` array? Can they only be used on the LP variant? Would there be any harm if we removed FW filename from them too? Maybe from other architectures too? It's a nice simplification not to have to repeat the same code everywhere, let alone saving a 100 bytes of memory from remove a dozen of strings :-)

**Review by RanderWang — DISMISSED** [review]

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 2 above)
Comment (libinyang) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:355]: @lyakh Good question. I talked with Bard when I made this patch and we decided to apply this patch only on the platforms which need patch, for safety. @plbossart @bardliao Do you have suggestion whether we should apply this patch on other platforms or only on this platform?

Comment (bardliao) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:355]: @libinyang I agree with @lyakh. We can remove all `.sof_fw_filename` in soc-acpi-intel-tgl-match.c now, and review other platforms latter. In theory, we can remove them all except for special cases.

**Comment by libinyang** [comment]
patch is updated based on @lyakh 's suggestion

File: sound/soc/codecs/rt715-sdca.c [hunk 3]
```diff
@@ -300,7 +300,9 @@ static int rt715_sdca_put_volsw(struct snd_kcontrol *kcontrol,
 	.info = rt715_sdca_fu_info, \
 	.get = rt715_sdca_get_volsw, \
 	.put = rt715_sdca_put_volsw, \
-	.private_value = RT715_SDCA_FU_VALUE(reg_left, reg_right, xshift, \
+	//.private_value = RT715_SDCA_FU_VALUE(reg_left, reg_right, xshift, \
+	//				    xmax, xinvert)}
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: looks like a left-over

File: sound/soc/codecs/rt715-sdca.c [hunk 4]
```diff
@@ -300,7 +300,9 @@ static int rt715_sdca_put_volsw(struct snd_kcontrol *kcontrol,
...
+	.private_value = SOC_DOUBLE_R_VALUE(reg_left, reg_right, xshift, \
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: the whole hunk doesn't belong to this PR, right?

**Review by lyakh — CHANGES_REQUESTED** [review]

File: sound/soc/codecs/rt715-sdca.c (hunk 4 above)
Comment (libinyang) [reply sound/soc/codecs/rt715-sdca.c]: > the whole hunk doesn't belong to this PR, right?

@lyakh Thanks. Done

**Review by bardliao — APPROVED** [review]
LGTM

**Review by lyakh — APPROVED** [review]

**Comment by plbossart** [comment]
@libinyang Why would we have a different solution for firmware names depending on whether we use the SoundWire interface or not?
And why do it for TGL and SoundWire only? 

This is not helpful IMHO, this just adds confusion and misaligned changes between platforms.

**Comment by libinyang** [comment]
> @libinyang Why would we have a different solution for firmware names depending on whether we use the SoundWire interface or not?
> And why do it for TGL and SoundWire only?
> 
> This is not helpful IMHO, this just adds confusion and misaligned changes between platforms.

@plbossart It should apply not only to TGL, but also to other platforms. Currently only TGL met such situation. TGL has 2 kinds of FW, sof-tgl.ri and sof-tgl-h.ri. It depends on the PCI ID (tgl-h and tgl-u are using different pciid). And it doesn't depends on the SDW link topoloy, as tgl-h and tgl-u can have the same link topology. So we can't tell which fw should be used in struct snd_soc_acpi_mach.

**Comment by plbossart** [comment]
Ah, that's interesting, thanks @libinyang for adding more context.
Indeed we need to think about this, it could very well be that the machine descriptor has reached its limits.
